### DISCUSSION TITLE
How to extract the mock out to a module ?

### DISCUSSION THREAD

**Discussion by ygqygq2** [vitest-dev/vitest#5454] [Q&A] [answered]
`globalMocks.ts`:

```
import { vi } from 'vitest';

export function mockLogger() {
  return {
    logger: {
      trace: vi.fn(),
      debug: vi.fn(),
      info: vi.fn(),
      warn: vi.fn(),
      error: vi.fn(),
    },
  };
}
```

Then in a test file use it: 
```
vi.mock('@/extension', mockLogger);
```
But got `ReferenceError: Cannot access '__vi_import_1__' before initialization`.

Can I only use this in each test file ?

```
vi.mock('@/extension', () => ({
  logger: {
    trace: vi.fn(),
    debug: vi.fn(),
    info: vi.fn(),
    warn: vi.fn(),
    error: vi.fn(),
  },
}));
```

**Comment by hi-ogawa** [comment]
Can you provide a reproduction? I'm guessing that you have something like this:

```ts
//// globalMocks.ts:
import { vi } from 'vitest';
export function mockLogger() { ... }

//// some-file.test.ts
import { vi } from "vitest"
import { mockLogger } from "./globalMocks.ts"

vi.mock("@/extension", mockLogger);
```

But, as described in https://vitest.dev/api/vi.html#vi-mock saying "The call to vi.mock is hoisted", this is transformed to roughly something like this:

```ts
//// some-file.test.ts
import { vi } from "vitest";
vi.mock("@/extension", __vi_import_1__.mockLogger);

const __vi_import_1__ = await import("./globalMocks.ts")
```

I don't remember the intricate detail of this hoisting, but probably this would work though it's fairly ugly...

```ts
//// some-file.test.ts
import { vi } from "vitest"

vi.mock("@/extension", () => import("./globalMocks.ts").then(v => v.mockLogger));
```

On a related note, a different approach is to use [`setupFiles`](https://vitest.dev/config/#setupfiles) to setup `vi.mock('@/extension', () => {...})`, but this will setup mock for all the test files, so it might not be what you actually want to do.

**Reply by ygqygq2** [reply]
https://github.com/ygqygq2/turbo-console-log/tree/main/src/test/unit/commands
The following files are used this:

```
vi.mock('@/extension', () => ({
  logger: {
    trace: vi.fn(),
    debug: vi.fn(),
    info: vi.fn(),
    warn: vi.fn(),
    error: vi.fn(),
  },
}));
```
So, if there's more of the same code, I want to extract it, but as you say, I may not need them in all test files.

**Reply by hi-ogawa** [reply, marked as answer]
Thanks for the reference. Yeah, I see the convenience of adding `vi.mock("@/extension", mockLogger)` to only required files, but it might not be possible currently.

I think you need to do dynamic import in one way or another. Maybe another possibility is:

```ts
//// some-file.test.ts
import { vi } from "vitest"
const { mockLogger } = await vi.hoisted(() => import("./globalMocks.ts"));
vi.mock("@/extension", mockLogger);
```

**Reply by ygqygq2** [reply]
Thanks. Vscode extension tsconfig.json now can support `ESNext` module, so I can use it like this:

```
const { mockLogger } = await vi.hoisted(() => import("@/test/globalMocks"));
vi.mock("@/extension", mockLogger);
```
